scripts/lldb/firestore.py

```diff
@@ -0,0 +1,250 @@
+
+  def get_child_index(self, name):
+    return self.delegate().GetIndexOfChildWithName(name)
+
+  def get_child_at_index(self, index):
+    return self.delegate().GetChildAtIndex(index)
+
+  def update(self):
+    pass
+
+
+# Abseil
+
+class AbseilOptional_SynthProvider(object):
+  """A synthetic child provider that hides the internals of absl::optional.
+  """
+
+  def __init__(self, value, params):
+    self.value = value
+    self.engaged = None
+    self.data = None
+
+  def update(self):
+    # Unwrap all the internal optional_data and similar types
+    value = self.value
+    while True:
+      if value.GetNumChildren() <= 0:
+        break
+
+      child = value.GetChildAtIndex(0)
+      if not child.IsValid():
+        break
+
+      if 'optional_internal' not in child.GetType().GetName():
+        break
+
+      value = child
+
+    # value should now point to the innermost absl::optional container type.
+    self.engaged = value.GetChildMemberWithName('engaged_')
+
+    if self.has_children():
+      self.data = value.GetChildMemberWithName('data_')
+
+    else:
+      self.data = None
+
+  def has_children(self):
+    return self.engaged.GetValueAsUnsigned(0) != 0
+
+  def num_children(self):
+    return 2 if self.has_children() else 1
+
+  def get_child_index(self, name):
+    if name == 'engaged_':
+      return 0
+    if name == 'data_':
+      return 1
+    return -1
+
+  def get_child_at_index(self, index):
+    if index == 0:
+      return self.engaged
+    if index == 1:
+      return self.data
+
+
+def AbseilOptional_SummaryProvider(value, params):
+  # Operates on the synthetic children above, calling has_children.
+  return 'engaged={0}'.format(format_bool(value.MightHaveChildren()))
+
+
+# model
+
+class DatabaseId_SynthProvider(ForwardingSynthProvider):
+  """Makes DatabaseId behave as if `*rep_` were inline."""
+  def delegate(self):
+    return deref_shared(self.value.GetChildMemberWithName('rep_'))
+
+
+def DatabaseId_SummaryProvider(value, params):
+  # Operates on the result of the SynthProvider; value is *rep_.
+  parts = [
+      get_string(value.GetChildMemberWithName('project_id')),
+      get_string(value.GetChildMemberWithName('database_id'))
+  ]
+  return format_string('/'.join(parts))
+
+
+def DocumentKey_SummaryProvider(value, params):
+  """Summarizes DocumentKey as if path_->segments_ were inline and a single
+  string."""
+  return deref_shared(value.GetChildMemberWithName('path_')).GetSummary()
+
+
+def ResourcePath_SummaryProvider(value, params):
+  """Summarizes ResourcePath as if segments_ is a single string."""
+
+  segments = value.GetChildMemberWithName('segments_')
+  count = segments.GetNumChildren()
+
+  segment_text = []
+  for i in range(0, count):
+    child = segments.GetChildAtIndex(i)
+    segment_text.append(get_string(child))
+
+  text = format_string('/'.join(segment_text))
+  return text
+
+
+# api
+
+def DocumentReference_SummaryProvider(value, params):
```

Comment:
Nit: these few summaries lack a doc comment. If the comments are a pain to write, I think just an example is sufficient.

Reply:
Given that these were one-liners indicating that the summary was derived from a single field I considered these fairly self-explanatory.

I've added comments but I suspect I'm not quite capturing what you're after.

Reply:
I like the new comments, thanks for adding them.